src/common/range.rs

```diff
@@ -141,278 +146,454 @@ impl ByteRangeSpec {
+    pub fn to_range_bounds(&self, len: u64) -> Option<impl RangeBounds<u64>> {
         // If the full length is zero, there is no satisfiable end-inclusive range.
-        if full_length == 0 {
+        if len == 0 {
             return None;
         }
         match self {
-            &ByteRangeSpec::FromTo(from, to) => {
-                if from < full_length && from <= to {
-                    Some((from, ::std::cmp::min(to, full_length - 1)))
-                } else {
-                    None
-                }
-            },
-            &ByteRangeSpec::AllFrom(from) => {
-                if from < full_length {
-                    Some((from, full_length - 1))
+            ByteRangeSpec::FromTo(first, last) => {
+                // If the index of the first byte is greater or equal to
+                // the length, the range is not satisfiable
+                if *first >= len {
+                    return None;
+                };
+                // If the index of the last byte is beyond the end,
+                // take the last byte index instead
+                let last = if *last >= len {
+                    Bound::Excluded(len)
+                // If the index of the first byte is after the index
+                // of the last byte, the range is not satisfiable
+                } else if first > last {
+                    return None;
+                // In this case, 0 <= first <= last <= len, so we can
+                // use last directly
                 } else {
-                    None
+                    Bound::Included(*last)
+                };
+                Some((Bound::Included(*first), last))
+            }
+            ByteRangeSpec::AllFrom(first) => {
+                if *first >= len {
+                    return None;
                 }
-            },
-            &ByteRangeSpec::Last(last) => {
-                if last > 0 {
-                    // From the RFC: If the selected representation is shorter
-                    // than the specified suffix-length,
-                    // the entire representation is used.
-                    if last > full_length {
-                        Some((0, full_length - 1))
-                    } else {
-                        Some((full_length - last, full_length - 1))
-                    }
+                Some((Bound::Included(*first), Bound::Excluded(len)))
+            }
+            ByteRangeSpec::Last(suf_len) => {
+                // If the suffix length is bigger than the resource length
+                // itself, return the whole resource.
+                let start = if suf_len.get() > len {
+                    0
+                // else, subtract the suffix length from the resource length
+                // and take as the first byte
                 } else {
-                    None
-                }
+                    len - suf_len.get()
+                };
+                Some((Bound::Included(start), Bound::Excluded(len)))
```

Comment:
TIL, I didn't know about these methods so far. I'm not quite sure whether these are any better though, because the usage of these relatively obscure features might make it harder for people to understand what's going on.

Reply:
My perspective on this:
- this is a standard library method
- it's easy to look up and doesn't do anything complicated
- what you're doing is essentially reimplementing `saturating_sub`
So I think it's better to use the method, but I see your point too! Another thing is that you can toss a comment explaining what you're doing to make it more obvious.

Reply:
included in the current version, resolving